gnocchi/amqp1d.py

```diff
@@ -0,0 +1,221 @@
+                                         resource_id,
+                                         self.conf.amqp1d.creator,
+                                         original_resource_id=host_id,
+                                         host=self._hosts[host_id])
+        except indexer.ResourceAlreadyExists:
+            pass
+        else:
+            LOG.info("Created resource for %s", resource_id)
+
+        return self.indexer.get_resource(self.conf.amqp1d.resource_type,
+                                         resource_id, with_metrics=True)
+
+    def _ensure_resource_type_exists(self):
+        try:
+            self.resource_type = self.indexer.get_resource_type(
+                self.conf.amqp1d.resource_type)
+        except indexer.NoSuchResourceType:
+            try:
+                mgr = self.indexer.get_resource_type_schema()
+                rtype = mgr.resource_type_from_dict(
+                    self.conf.amqp1d.resource_type, {
+                        "host": {"type": "string", "required": True,
+                                 "min_length": 0, "max_length": 255},
+                    }, "creating")
+                self.indexer.create_resource_type(rtype)
+            except indexer.ResourceTypeAlreadyExists:
+                LOG.debug("Resource type %s already exists",
+                          self.conf.amqp1d.resource_type)
+            else:
+                LOG.info("Created resource type %s",
+                         self.conf.amqp1d.resource_type)
+                self.resource_type = self.indexer.get_resource_type(
+                    self.conf.amqp1d.resource_type)
+        else:
+            LOG.info("Found resource type %s",
+                     self.conf.amqp1d.resource_type)
+
+
+class CollectdFormatHandler(object):
+    def __init__(self, processor):
+        self.processor = processor
+
+    @staticmethod
+    def _serialize_identifier(index, message):
+        """Based of FORMAT_VL from collectd/src/daemon/common.h.
+
+        The biggest difference is that we don't prepend the host and append the
+        index of the value, and don't use slash.
+
+        """
+        suffix = ("-%s" % message["dsnames"][index]
+                  if len(message["dsnames"]) > 1 else "")
+        return (message["plugin"] + ("-" + message["plugin_instance"]
+                                     if message["plugin_instance"] else "")
+                + "@"
+                + message["type"] + ("-" + message["type_instance"]
+                                     if message["type_instance"] else "")
+                + suffix)
+
+    def on_message(self, event):
+        json_message = ujson.loads(event.message.body)
+        for message in json_message:
+            for index, value in enumerate(message["values"]):
+                try:
+                    name = self._serialize_identifier(index, message)
+                    self.processor.treat_metric(message["host"], name, value)
```

Comment:
This should probably be done in batch.
Calling `utcnow()` each time in `treat_metric` is going to make things slow.

Reply:
done